### ISSUE TITLE
Access resource plugin definition from an authentication plugin

### ISSUE THREAD

**Issue by hussainweb** [RESTful-Drupal/restful#939]
I am trying to write an authentication plugin for OAuth2 support and need to access resource's plugin definition. The idea is to allow defining the oauth2 server and scope required to access a particular resource.

OAuth2 server supports RESTful 7.x-1.x but needs variables to specify server and scope (see #111, and https://www.drupal.org/node/2451303). This is simple if you only have one server and scope. I imagine more than one OAuth2 server would be rare but more than one scope could be common. Anyway, variables don't seem like the best way to handle the issue anyway.

My problem now is that I cannot access the plugin definition from authentication plugin at all. I only get a Request object and using that (and the global restful controller using `restful()`), I can recreate the plugin and get the definition but that is a lot of code for something that could be done simply. Also, it might involve recreating the plugin just to get the definition.

Am I missing something here?

**Comment by hussainweb** [comment]
I started working with the global way to access plugins and noticed another problem. This is my code:

```
$resource = restful()->getResourceManager()->negotiate();
$plugin = $resource->getPluginDefinition();
```

This works but fails for discovery endpoint, because the above returns the definition for discovery resource. This is called by the discovery plugin's view logic that checks for user's access to each plugin.

Ideally, there should be a way to pass in the context to the authentication plugin which contains the resource plugin being checked. In case of a direct request, it will be the same plugin returned as above. In case of discovery plugin (and maybe other similar cases), it will be different.

**Comment by e0ipso** [comment]
Going the other way around is easier, since given a plugin definition and a request getting the authentication plugin is piece of cake.

As you point out the best way to do this is to use the _"service container"_ to get the resource manager and the resource plugin from there.

> This works but fails for discovery endpoint, because the above returns the definition for discovery resource

I'm not sure what you mean by _discovery resource_, all resources should have the discovery and data capabilities. That is triggered based on the HTTP verb (OPTIONS/GET).

**Comment by hussainweb** [comment]
@e0ipso, thanks for your response.

Two lines to load the plugin is fair, but does this not instantiate the plugin again? I hope there are no side effects in doing this. I am only interested in the definition. This is how the plugin definition currently looks:

```
/**
 * Users endpoint.
 *
 * @Resource(
 *   name = "user:1.0",
 *   resource = "user",
 *   label = "Users",
 *   description = "List, get, and update users on the site.",
 *   authenticationTypes = {
 *     "oauth2_auth"
 *   },
 *   oauth2Server = "oauth2_server",
 *   oauth2Scope = "profile",
 *   dataProvider = {
 *     "entityType": "user",
 *     "bundles": {
 *       "user"
 *     },
 *   },
 *   majorVersion = 1,
 *   minorVersion = 0
 * )
 */
```

As you can see, I am interested in `oauth2Server` and `oauth2Scope` in the authentication plugin. I can now get that using the service container like you said but I was hoping for a simpler method.

About the problem with discovery: I am talking about the discovery plugin at example.com/api/. This endpoint returns information on all discoverable resources if they can be accessed by the requesting user. It calls the authentication process on each of the providers to verify if it can be used and only returns information on those that can be accessed. But since my authenticate method relies on the global state to get the plugin, it only sees the discovery plugin and not the plugin discovery resource is trying to check. Does this make sense?

**Comment by hussainweb** [comment]
Okay, on debugging further, my code does not really work because of the `restful()->getResourceManager()->negotiate()` call. This results in an infinite loop because `negotiate()` tries to authenticate again which results in the `negotiate` call again. :(

**Comment by hussainweb** [comment]
This is closed with #940
